src/bin/book-test-instance.py

```diff
@@ -0,0 +1,409 @@
+
+    dated_list.sort(key=lambda i: i[1])
+
+    if not dated_list[0][0]:
+        raise Exception('No available instance could be found.')
+
+    return dated_list[0][0]
+
+
+def get_instances():
+    """
+    Uses cloud function also used on https://greenpeace.github.io/planet4-test-swarm/
+
+    Return swarm instances
+    """
+    return api_query(SWARM_API)
+
+
+def get_instance_last_commit_date(instance):
+    """
+    Return last commit date for an instance repo
+    """
+    commit = api_query(
+        GITHUB_API + '/repos/' + INSTANCE_REPO_PREFIX + instance + '/commits/main',
+        {'Accept': 'application/vnd.github.v3+json'}
+    )
+
+    return commit['commit']['committer']['date']
+
+
+"""
+
+API stuff
+
+"""
+
+
+def api_query(url, headers={'Accept': 'application/json'}, auth=None):
+    """
+    Queries API
+    - fails on error
+    - return json
+    - use cache if requested
+    """
+
+    vprint('GET {0}'.format(url))
+
+    cache_file = '/tmp/{0}.cache'.format(hashlib.md5(url.encode()).hexdigest())
+    if use_request_cache and os.path.isfile(cache_file):
+        vprint('Using cache for ' + url)
+        return json.load(open(cache_file))
+
+    response = requests.get(url, headers=headers, auth=auth)
+    api_failed(response, url)
+
+    content = response.json()
+
+    if use_request_cache:
+        vprint(url, ' -> ', cache_file)
+
+    json.dump(content, open(cache_file, "w"))  # write anyway
+    return content
+
+
+def api_failed(response, endpoint, exit_on_error=True):
+    """
+    Check if api request failed
+
+    Can raise exception
+    """
+
+    if 200 <= response.status_code < 300:
+        return False
+
+    vprint('API call failed')
+    vprint(response.status_code, response.headers, response.text)
+    if exit_on_error:
+        raise Exception("Status code {0} calling {1}".format(
+            response.status_code, endpoint))
+    return True
+
+
+def get_jira_auth():
+    """
+    Based on env variables
+
+    https://developer.atlassian.com/server/jira/platform/oauth/
+    """
+
+    if os.getenv('JIRA_CLIENT_KEY'):
+        if os.getenv('JIRA_PRIVATE_KEY'):
+            key = os.getenv('JIRA_PRIVATE_KEY')
+        else:
+            key = None
+
+        return {
+            'type': 'oauth',
+            'auth': OAuth1(
+                os.getenv('JIRA_CLIENT_KEY'),
+                resource_owner_key=os.getenv('JIRA_OAUTH_TOKEN'),
+                resource_owner_secret=os.getenv('JIRA_OAUTH_SECRET'),
+                rsa_key=key,
+                signature_method=SIGNATURE_RSA,
+            )
+        }
+
+    return {
+        'type': 'basic',
+        'auth': HTTPBasicAuth(os.getenv('JIRA_USERNAME'), os.getenv('JIRA_PASSWORD'))
+    } if os.getenv('JIRA_USERNAME') else {}
+
+
+def save_results(results, filename='booking-results.json'):
+    with open(filename, 'w') as results_file:
+        json.dump(results, results_file, indent=2)
+
+
+"""
+
+Main
+
+"""
+
+if __name__ == '__main__':
```

Comment:
It feels safer to invert this if statement, since it uses indentation to detect what is considered in the body. If somewhere along the indentation of a line changes, it's hard to notice that this breaks the if statement body.

I guess the point of these is to prevent executing the code if the file is imported as a module. I guess we never intend to do that? In that case we could have it throw an error if the name is not `__main__`.

Something like 
```python
if __name__ !== '__main__':
        raise Exception('Trying to load main file as module.')
        
// ... happy non-indented code

Reply:
This is indeed a way to distinguish whether the module file is called as a script or imported as a module. For instance, the functions in this file can be imported if we want to. Reversing the statement and raising an exception prevents us from re-using code.

It's also a pretty standard line of code for python scripts:
https://docs.python.org/3/library/__main__.html

Reply:
Right, also if like you mention we'll move the functions out, then this if-statement will be near the top of the file and will contain less complex code inside :ok_hand: